File: aller-scan-api/app/routes/resteraunts.py
from typing import Any

from beanie.exceptions import RevisionIdWasChanged
from fastapi import APIRouter, HTTPException, Query, status
from pymongo import errors

from .. import schemas, models

router = APIRouter()

@router.post("/", response_model=schemas.Resteraunt, status_code=status.HTTP_201_CREATED)
async def create_resteraunt(resteraunt_in: schemas.ResterauntCreate):
    existing_resteraunt = await models.Resteraunt.get(resteraunt_in.google_maps_id)

    if existing_resteraunt is not None:
        raise HTTPException(status_code=400, detail="Resteraunt already exist")

    resteraunt = models.Resteraunt(
        id=resteraunt_in.google_maps_id,
        added_by=resteraunt_in.added_by,
        opening_times=resteraunt_in.opening_times,
        location=resteraunt_in.location,
        website_url=resteraunt_in.website_url,
        menu_items=resteraunt_in.menu_items,
        properties=resteraunt_in.properties,
    )

    try:
        await resteraunt.create()
        return resteraunt
    except errors.DuplicateKeyError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Resteraunt with that google_maps_id already exist"
        )

@router.put("/{resteraunt_id}", response_model=schemas.Resteraunt)
async def update_resteraunt(resteraunt_id: str, update: schemas.ResterauntUpdate):
    resteraunt = await models.Resteraunt.get(resteraunt_id)

    if resteraunt is None:
        raise HTTPException(status_code=404, detail="Resteraunt not found")

    updated_data = update.model_dump(exclude=True)

    updated_resteraunt = resteraunt.model_copy(update=updated_data)

    try:
        await updated_resteraunt.save()
        return updated_resteraunt
    except (errors.DuplicateKeyError, RevisionIdWasChanged):
        raise HTTPException(status_code=400, detail="Resteraunt with that google id is already exist")
    

@router.delete("/{resteraunt_id}", response_model=schemas.Resteraunt)
async def delete_resteraunt(reseraunt_id: str):
    resteraunt = await models.Resteraunt.get(reseraunt_id)

    if resteraunt is None:
         raise HTTPException(status_code=404, detail="resteraunt not found")

    await resteraunt.delete()

    return resteraunt


@router.get("/nearby", response_model=list[schemas.Resteraunt])
async def get_nearby(
    latitude: float = Query(..., ge=-90, le=90, description="User latitude"),
    longitude: float = Query(..., ge=-180, le=180, description="User longitude"),
    radius_meters: float = Query(1000.0, gt=0, description="radius in meters")):
        close_restaurants = await models.Resteraunt.find(
            {
                "location.coordinates": {
                    "$near": {
                        "$geometry": {
                            "type": "Point",
                            "coordinates": [longitude, latitude]
                        },
                        "$maxDistance": radius_meters
                    }
                }
            }
        ).to_list()

        if len(close_restaurants) == 0:
            raise HTTPException(status_code=404, detail="Close resteraunts were not found, please increase the radius")

        return close_restaurants

@router.get("/{resteraunt_id}", response_model=schemas.Resteraunt)
async def get_by_id(resteraunt_id: str):
    resteraunt = await models.Resteraunt.get(resteraunt_id)

    if resteraunt is None:
        raise HTTPException(status_code=404, detail="resteraunt not found")

    return resteraunt

File: aller-scan/src/api/resteraunts.ts
import { ALLER_SCAN_API_BASE_URL } from "./config";
import { ApiError, get, put, post, del } from "./httpClient";

export interface Resteraunt {
    id: string
    resteraunt_name: string,
    added_by: string
    opening_times: string[],
    location: Location
    website_url?: string
    menu_items: MenuItem[]
    properties: Record<string, boolean>
}

export interface Location {
    full_address: string
    coordinates: number[]
}

export interface MenuItem {
    item_name: string
    category: string
    ingredients: string[]
    allergens: string[]
}

export interface ResterauntCreateInput {
    id: string
    added_by: string
    resteraunt_name: string
    opening_times: string[]
    location: Location
    menu_items: MenuItem[]
    website_url?: string
    properties?: Record<string, boolean>
}

export interface ResterauntUpdateInput {
    opening_times: string[]
    menu_items: MenuItem[]
    website_url?: string
    properties?: Record<string, boolean>
}

export async function getResteraunt(id: string): Promise<Resteraunt | null> {
    try {
        const resteraunt = await get<Resteraunt>(`${ALLER_SCAN_API_BASE_URL}/resteraunts/${id}`);
        return resteraunt;
    } catch (err) {
        if (err instanceof ApiError && err.status === 404) return null;
        throw err;
    }
}

export async function createResteraunt(data: ResterauntCreateInput): Promise<Resteraunt> {
    return post<Resteraunt>(`${ALLER_SCAN_API_BASE_URL}/resteraunts/`, { json: data });
}

export async function updateResteraunt(id: string, data: ResterauntUpdateInput): Promise<Resteraunt> {
    return put<Resteraunt>(`${ALLER_SCAN_API_BASE_URL}/resteraunts/${id}`, { json: data });
}

export async function deleteResteraunt(id: string): Promise<Resteraunt> {
    return del<Resteraunt>(`${ALLER_SCAN_API_BASE_URL}/resteraunts/${id}`);
}

export async function getNearbyResteraunts(
    latitude: number,
    longitude: number,
    radiusMeters?: number
): Promise<Resteraunt[]> {
    try {
        return await get<Resteraunt[]>(`${ALLER_SCAN_API_BASE_URL}/resteraunts/nearby`, {
            query: { latitude, longitude, radius_meters: radiusMeters },
        });
    } catch (err) {
        if (err instanceof ApiError && err.status === 404) return [];
        throw err;
    }
}

